ui/main.cpp

```diff
@@ -232,6 +238,57 @@ auto SetupFlags(int argc, char **argv)
     return absl::ParseCommandLine(argc, argv);
 }
 
+//--------------------------------------------------------------------------------------------------
+std::optional<std::filesystem::path> GetTestSavePath(std::string_view filename)
+{
+    if (absl::GetFlag(FLAGS_test_output).empty())
+    {
+        return std::nullopt;
+    }
+    std::filesystem::path output_dir(absl::GetFlag(FLAGS_test_output));
+    std::filesystem::create_directories(output_dir);
+    if (!std::filesystem::is_directory(output_dir))
+    {
+        return std::nullopt;
+    }
+    std::string full_filename = absl::GetFlag(FLAGS_test_prefix);
+    full_filename += filename;
+    return output_dir / full_filename;
+}
+
+//--------------------------------------------------------------------------------------------------
+bool ExecuteScenario(std::string_view scenario, MainWindow *main_window)
+{
+    if (scenario == "exit-after-load")
+    {
+        QObject::connect(main_window, &MainWindow::FileLoaded, main_window, &MainWindow::close);
+        return true;
+    }
+
+    if (scenario == "screenshot")
+    {
+        auto savepath = GetTestSavePath("MainWindow.png");
+        if (!savepath)
+        {
+            qDebug() << "Invalid screenshot path";
+            return false;
+        }
+        auto func = [main_window, savepath]() {
+            QPixmap pixmap(main_window->size());
+            main_window->render(&pixmap);
+            pixmap.save(QString::fromStdString(savepath->string()));
+            main_window->close();
+        };
+        QObject::connect(main_window, &MainWindow::FileLoaded, main_window, [func, main_window]() {
+            QTimer::singleShot(kScreenshotDelay, main_window, func);
+        });
```

Comment:
nit: consider renaming `func` to something more self-documenting like `take_screenshot`
```suggestion
        auto take_screenshot = [main_window, savepath]() {
            QPixmap pixmap(main_window->size());
            main_window->render(&pixmap);
            pixmap.save(QString::fromStdString(savepath->string()));
            main_window->close();
        };
        QObject::connect(main_window, &MainWindow::FileLoaded, main_window, [take_screenshot, main_window]() {
            QTimer::singleShot(kScreenshotDelay, main_window, take_screenshot);
        });
```